New section for algorithms and category maps

vanessavmac opened the pull request:
## Summary

Add a new algorithms overview. Allow users to view the algorithm URI from the identification summary.

### Related Issues

Closes #683 

### Screenshots

Algorithm details and overview
![Screenshot 2025-04-10 at 10 33 05 AM](https://github.com/user-attachments/assets/dfefb738-f61d-4f5d-aee8-4364d8f33d6a)
![Screenshot 2025-04-10 at 10 33 18 AM](https://github.com/user-attachments/assets/7defcd9f-592a-41fb-8041-3fb5c8e900ca)
Pipeline links to algorithms
![Screenshot 2025-04-10 at 10 33 35 AM](https://github.com/user-attachments/assets/adebcb96-f347-4bd2-9951-9138b24ea8d1)
Identification summary links to algorithm
![Screenshot 2025-04-10 at 10 33 57 AM](https://github.com/user-attachments/assets/d0771a90-237d-4558-9da2-b9109ab2766e)

mihow commented:
This is looking great Vanessa! FYI, Anna just merged a big change to the project overview change. So make sure to bring this branch up-to-date with `main`. But you may need her help as well. She added routes to detail pages for all of the tabbed views, so now you can link to a specific algorithm modal!

annavik commented:
Hi Vanessa! Great work on this!

Yes as Michael said we have made some updates to the overview! Please sync with main and try include this as a page under our new section "Processing". Happy to jump on a call and provide more details, give me a ping! :)

Some early thoughts on the UI:
- I think the field "URI" is a bit long to present in UI this way. What do you think about skip in table view and present as an external link button in detail view?
- I think the field "Category Map ID" requires some more context. What do you think about skip in table view and present as an external link (?) in detail view, with some more information?
- What do you think about including algorithm description in table view?
- Sorting seems to not be working, although presented as an available option in UI.
- The external link button in identifications tab looks a bit misaligned and is also missing some padding.
- I think we should try integrate this work with pipeline details, maybe simplify the algorithm section a bit in this view and link to algorithm details, to make UI less fragmented.

annavik commented:
This is just a suggestion, but for pipeline details, I was thinking perhaps we could do something like this? 
<img width="1728" alt="Screenshot 2025-03-24 at 15 45 33" src="https://github.com/user-attachments/assets/3091e454-0c87-4f5c-9d4c-cc7289f25cc7" />

vanessavmac commented:
Thanks for the feedback @annavik ! I made some of the changes, a few I had some questions about:

> I think the field "URI" is a bit long to present in UI this way. What do you think about skip in table view and present as an external link button in detail view?
done

<img width="716" alt="Screenshot 2025-03-27 at 1 10 12 AM" src="https://github.com/user-attachments/assets/fc0ca1f9-b48a-456c-9421-2d9fa154507f" />
<img width="992" alt="Screenshot 2025-03-27 at 1 13 56 AM" src="https://github.com/user-attachments/assets/5f87a6f6-d645-4185-8fc7-29af2d520a30" />

> I think the field "Category Map ID" requires some more context. What do you think about skip in table view and present as an external link button (?) in detail view, with some more information?

I added this new section to the algorithm details called "Category Map Details." What do you think about this? I'm currently using a `MinimalNestedModelSerializer` for category maps to only provide the category map id and URI, but I can change this depending on what values we'd like to display in this section.

<img width="717" alt="Screenshot 2025-03-27 at 1 11 31 AM" src="https://github.com/user-attachments/assets/b7ca5dd3-be22-4881-8081-0f58af2a3127" />

> What do you think about including algorithm description in table view?

Yes I think that would be good, but main concern here is it might get long -- I added it to test so let me know what you think.

> Can we try streamline column titles and field labels to use "Sentence case" instead of "Title Case"?

Updated, hopefully I didn't miss any, will keep this in mind in future

> Sorting seems to not be working, although presented as an available option in UI.

I'm not sure how to solve this bug, I used the same component structure as the pipelines so I'm not sure why this is happening. Could we sync up on this?

> The external link button in identifications tab looks a bit misaligned and is also missing some margin.

Yes, is there a standard way/best practice I should be following for adding custom styling to components?

In particular, this is the component that would need to be styled
```
<Link to={identification.algorithm.uri}>
  <Tooltip content={identification.algorithm.description}>
    <div className={styles.details}>
      {identification.algorithm.name}
      <Icon
        type={IconType.ExternalLink}
        theme={IconTheme.Primary}
        size={16}
      />
    </div>
  </Tooltip>
</Link>
```

> I think we should try integrate this work with pipeline details, maybe simplify the algorithm section a bit in this view and link to algorithm details, to make UI less fragmented.

I agree, I think it's probably best to go over this in a call. I'll ping you!

vanessavmac commented:
TODOs
- See category map details (with a link on it) -- anna will send me a code snippet for a button
- sorting might be a backend issue (might just remove sort key from columns.tsx

> I think we should try integrate this work with pipeline details, maybe simplify the algorithm section a bit in this view and link to algorithm details, to make UI less fragmented.

talk with michael about this

mihow commented:
Some notes from discussion with Anna:
- Fix link to algorithm detail page from the occurrence prediction (currently links to the Algorithm URI instead)
- Link between Algorithms & Pipelines where possible
- Link to the Category Map API view
- Add number of species / categories to Algorithm view if possible
- Add fields allowed to be sorted on in the API view (ID, name, task_type, etc)

annavik commented:
> TODOs
> 
> * See category map details (with a link on it) -- anna will send me a code snippet for a button
> * sorting might be a backend issue (might just remove sort key from columns.tsx
> 
> > I think we should try integrate this work with pipeline details, maybe simplify the algorithm section a bit in this view and link to algorithm details, to make UI less fragmented.
> 
> talk with michael about this

Here is a code snipped for how to style an external HTML link as an icon button:

```
import { ExternalLinkIcon } from 'lucide-react'
import { buttonVariants } from 'nova-ui-kit'

<a
  className={buttonVariants({
    size: 'small',
    variant: 'outline',
  })}
  href="Some URL"
  rel="noreferrer"
  target="_blank"
>
  <span>Some label</span>
  <ExternalLinkIcon className="w-4 h-4 ml-2" />
</a>
```

vanessavmac commented:
> Link between Algorithms & Pipelines where possible

Let me know if there's anywhere I can add this -- I was thinking inside the Algorithms list in "Pipeline Details" but the table component doesn't allow customizing links inside the columns

> Add number of species / categories to Algorithm view if possible

What's the best way to retrieve number of species? For example, if I visit a category map API endpoint, do I just count the number of labels? Or the number of data elements? http://localhost:8000/api/v2/ml/labels/11/

> Add fields allowed to be sorted on in the API view (ID, name, task_type, etc)

This might be a backend bug, but all of the fields (Except updated at and sorted at) cannot be sorted

Aside: I can address all these points in a follow up PR since they might require more planning/investigating

annavik commented:
Thanks for the updates Vanessa! Here is a short summary of what I would like to do when we meet:
- Connect pipeline view with algorithm view (using links)
- Open all external links in new tabs
- If internal link, skip external link icon
- Present category API link similar to API link in occurrence detail view
- Limit description column width
- Reorganize the order of fields a bit

After this I think we can merge! :) Remaining items would be:
- Present number of categories
- Fix sorting

annavik reviewed (approved):
Thanks for these new views! 🥳 From my side, I think we can merge this!

vanessavmac commented:
Nice, thanks @annavik for all the help! @mihow Can we go ahead and merge?

mihow commented:
Thanks so much for bringing this over the finish line @vanessavmac and @annavik! I went ahead and added the category counts and sortable fields! The category count was a slightly new challenge since we are using a special Postgres ArrayField, but it works quite well now and is fast.

mihow commented:
@annavik the one thing that isn't working for me is clicking on the name of an algorithm in the identification details of an occurrence. Is that saved for the for #804? I see some code in this PR for the click action, so just checking.
![image](https://github.com/user-attachments/assets/874753e0-9ae0-4fa1-b8bb-0c284342e787)

annavik commented:
> @annavik the one thing that isn't working for me is clicking on the name of an algorithm in the identification details of an occurrence. Is that saved for the for #804? I see some code in this PR for the click action, so just checking. ![image](https://private-user-images.githubusercontent.com/158175/432630198-874753e0-9ae0-4fa1-b8bb-0c284342e787.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQzNTQwMzIsIm5iZiI6MTc0NDM1MzczMiwicGF0aCI6Ii8xNTgxNzUvNDMyNjMwMTk4LTg3NDc1M2UwLTlhZTAtNGZhMS1iOGJiLTBjMjg0MzQyZTc4Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxMVQwNjQyMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02YzRiY2M2N2IzMzA2ODY0Yjc1ZjNiYzc5OTMwNTgzYWQ5MTE1ZDUwMjY4ODllYWY2NWNiY2MwNDY2OGRhYmFjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.RM4UovtaFsuovRJOIaDCMLWO03nSCIJRMi5O9hUj2H8)

That we do intend to fix! I'll have a quick look! :)

annavik commented:
> > @annavik the one thing that isn't working for me is clicking on the name of an algorithm in the identification details of an occurrence. Is that saved for the for #804? I see some code in this PR for the click action, so just checking. ![image](https://private-user-images.githubusercontent.com/158175/432630198-874753e0-9ae0-4fa1-b8bb-0c284342e787.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQzNTQwMzIsIm5iZiI6MTc0NDM1MzczMiwicGF0aCI6Ii8xNTgxNzUvNDMyNjMwMTk4LTg3NDc1M2UwLTlhZTAtNGZhMS1iOGJiLTBjMjg0MzQyZTc4Ny5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxMVQwNjQyMTJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT02YzRiY2M2N2IzMzA2ODY0Yjc1ZjNiYzc5OTMwNTgzYWQ5MTE1ZDUwMjY4ODllYWY2NWNiY2MwNDY2OGRhYmFjJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.RM4UovtaFsuovRJOIaDCMLWO03nSCIJRMi5O9hUj2H8)
> 
> That we do intend to fix! I'll have a quick look! :)

Hmm this is still working for me, we did update the UI kit lib component with a new prop here, maybe you need to rebuild your container if your are testing locally?

https://github.com/user-attachments/assets/7d37aa06-327e-42cb-8abc-1ed5ededdecd

annavik commented:
> Thanks so much for bringing this over the finish line @vanessavmac and @annavik! I went ahead and added the category counts and sortable fields! The category count was a slightly new challenge since we are using a special Postgres ArrayField, but it works quite well now and is fast.
> 
> ![image](https://private-user-images.githubusercontent.com/158175/432558606-12674b39-dab5-4303-aeca-12a7aea4e798.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDQzNTQ1OTYsIm5iZiI6MTc0NDM1NDI5NiwicGF0aCI6Ii8xNTgxNzUvNDMyNTU4NjA2LTEyNjc0YjM5LWRhYjUtNDMwMy1hZWNhLTEyYTdhZWE0ZTc5OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjUwNDExJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI1MDQxMVQwNjUxMzZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1lMWRmODczMzhmNTQ1NGJhODAyODU4NmNmZTRkMjJlY2JjZmFlNGNjZDk5ODY5NGIyZTc5ZTJkMDZlYjI0ZThhJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.5dXkgWrEltWYXW_IzdMm25DViBGZk6WHuulCcUpV90w)

Thanks Michael! It's great to see the count 🙌 For the sorting I just pushed some small tweaks to streamline default order, etc.

annavik commented:
Also as added a few updates to processing services and pipeline views, to keep behavior consistent with links and things, even if not fully related to this PR. Ok I think that was the last polish!

mihow commented:
Thank you @annavik! I still couldn't get the link to work after running `yarn install` and `yarn build` but I went ahead and deployed, and it works on live! Soo sweet. And it's very slick how you get get from Processing Service details -> Pipeline -> Algorithm now!
